code_01_dataprep_bno.R

```diff
@@ -0,0 +1,204 @@
+
+#'bind rows combining data frames, and filter
+dataGrouped = bind_rows(dataGrouped) %>% 
+  filter(duration_minutes <= 240, effort_distance_km <= 5, effort_area_ha <= 500)
+
+#'assign present or not
+dataGrouped = mutate(dataGrouped, pres_abs = observation_count >= 1)
+  
+#### load covariates from raster files ####
+#'load elevation and crop to hills size, then mask by hills
+#' sf
+alt = raster::raster("Elevation/alt/")
+cr = raster::crop(alt, raster::extent(as(hills, "Spatial")))
+alt.hills = raster::mask(cr, as(hills, "Spatial"))
+
+#'load evi layers
+EVI.all = raster::stack("EVI/MOD13Q1_EVI_AllYears.tif")
+x11();raster::plot(EVI.all)
+#'scale later
+#'EVI.all = EVI.all*0.0001
+names(EVI.all) = paste("evi", month.abb, sep = ".")
+
+#'load 5 year evi change(?) this is currently 6 years
+evifiles = list.files("EVI", pattern = "_201", full.names = T)
+evifiles = evifiles[as.numeric(stringi::stri_sub(evifiles, -8, -5)) >= 2013]
+
+#'make stack
+EVI.yearly = raster::stack(as.list(evifiles))
+#'scale stack later
+#EVI.yearly = EVI.yearly*0.0001
+
+#'resample to elevation resolution
+EVI.all.resam <-  raster::resample(EVI.all, alt.hills,method ="bilinear")
+EVI.yearly.resam <- raster::resample(EVI.yearly, alt.hills, method="bilinear")
+
+#'get slope and aspect
+slope <- raster::terrain(alt.hills, opt = 'slope', unit='degrees', neighbors = 8)
+aspect <- raster::terrain(alt.hills, opt='aspect', unit='degrees', neighbors = 8)
+
+
+#### extracting raster values #####
+#'make raster list
+landscapeRasters = list(alt.hills, EVI.all.resam, slope, aspect)
+
+#'make dataGrouped sf again
+dataGrouped = st_as_sf(dataGrouped, coords = c("X", "Y")) %>% st_set_crs(4326)
+
+#'map extract across the list
+landscapeData = map(landscapeRasters, function(x){
+  raster::extract(x, dataGrouped)
+})
+#'set names
+names(landscapeData) = c("elevation","evi","slope","aspect")
+
+#'bind into dataframe
+landscapeData = map(landscapeData, as_data_frame) %>% bind_cols() %>% 
+  `names<-`(c("elevation", paste("evi", month.abb, sep = "."), "slope", "aspect"))
+
+#'join with ebird data
+dataCovar = bind_cols(dataGrouped, landscapeData)
+
+#'remove rasters
+rm(alt, alt.hills, aspect, cr, EVI.all, EVI.yearly, EVI.yearly.resam, EVI.all.resam, slope)
+gc()
+
+#### mean time, distance, and number of observers ####
+
+dataCovar = dropGeometry(dataCovar)
+
+dataCovar = dataCovar %>% 
+  left_join(dataCovar %>% 
+      group_by(locality_id) %>% 
+      summarise_at(vars(duration_minutes, effort_distance_km, number_observers),
+                   funs(sum.no.na)) %>% 
+      left_join(count(dataCovar, locality_id)) %>% 
+      mutate_at(vars(duration_minutes, effort_distance_km, number_observers),
+                funs(./n)) %>% 
+        `names<-`(c("locality_id", "meanSampleTime", "meanSampleDist", "meanNObservers", "totalVisits"))
+  )
```

Comment:
There was a write_csv code which I don't see here, but was in one of your other files. I keep getting an error saying encode element.. something list (basically fails to write more than one line).

Reply:
This is the exact error: Error in write.table(dataOccuFin, "D:\\Occupancy Modeling\\Scripts\\Occupancy Scripts\\DataForOccuMod.csv",  : 
  unimplemented type 'list' in 'EncodeElement'

Reply:
I'll run it again and insert checks at each point - the write_csv is in code02*.r